metadata_layer.py

```diff
@@ -943,7 +943,7 @@ def from_edition(cls, edition):
         for contribution in edition.contributions:
             contributor = ContributorData.from_contribution(contribution)
             contributors.append(contributor)
-        else:
+        if not edition.contributions:
```

Comment:
This is the main one. That `else` clause was always being triggered because the `for` loop didn't break. In many cases this caused two `ContributorData` objects to be created for the same person: one with AUTHOR_ROLE (the code in the `for` loop) and one with PRIMARY_AUTHOR_ROLE (the code in the `else` statement).